ext/pdo_odbc/odbc_driver.c

```diff
@@ -590,30 +599,56 @@ static int pdo_odbc_handle_factory(pdo_dbh_t *dbh, zval *driver_options) /* {{{
 		/* Force UID and PWD to be set in the DSN */
 		if (dbh->username && *dbh->username && !strstr(dbh->data_source, "uid")
 				&& !strstr(dbh->data_source, "UID")) {
-			char uid[256], pwd[256];
-			/* XXX: Check for truncation by odbc_quote */
-			if (!is_odbc_quoted(dbh->username) && should_odbc_quote(dbh->username)) {
-				odbc_quote(uid, dbh->username, 256);
+			char *uid = NULL, *pwd = NULL;
+			bool should_quote_uid = !is_odbc_quoted(dbh->username) && should_odbc_quote(dbh->username);
+			bool should_quote_pwd = !is_odbc_quoted(dbh->password) && should_odbc_quote(dbh->password);
+			bool connection_string_fail = false;
+			if (should_quote_uid) {
+				size_t estimated_length = estimate_odbc_quote_length(dbh->username);
+				uid = emalloc(estimated_length);
+				if (!uid) {
+					connection_string_fail = true;
+					goto connection_string_fail;
+				}
+				odbc_quote(uid, dbh->username, estimated_length);
 			} else {
-				strlcpy(uid, dbh->username, 256);
+				uid = dbh->username;
 			}
 			/* XXX: Do we check if password is null? */
-			if (!is_odbc_quoted(dbh->password) && should_odbc_quote(dbh->password)) {
-				odbc_quote(pwd, dbh->password, 256);
+			if (should_quote_pwd) {
+				size_t estimated_length = estimate_odbc_quote_length(dbh->password);
+				pwd = emalloc(estimated_length);
+				if (!uid) {
+					connection_string_fail = true;
+					goto connection_string_fail;
+				}
+				odbc_quote(pwd, dbh->password, estimated_length);
 			} else {
-				strlcpy(pwd, dbh->password, 256);
+				pwd = dbh->password;
 			}
 			size_t new_dsn_size = strlen(dbh->data_source)
 				+ strlen(uid) + strlen(pwd)
 				+ strlen(";UID=;PWD=") + 1;
 			char *dsn = pemalloc(new_dsn_size, dbh->is_persistent);
 			if (dsn == NULL) {
+				connection_string_fail = true;
 				/* XXX: Do we inform the caller? */
-				goto fail;
+				goto connection_string_fail;
 			}
 			snprintf(dsn, new_dsn_size, "%s;UID=%s;PWD=%s", dbh->data_source, uid, pwd);
 			pefree((char*)dbh->data_source, dbh->is_persistent);
 			dbh->data_source = dsn;
+			/* Convoluted label to handle freeing in this scope */
+connection_string_fail:
```

Comment:
Yeah, that's how we usually deal with deallocation needs. It may make sense to use `alloca()` instead of `emalloc()` for performance reasons, but that would not avoid the manual cleanup, for portability reasons; see https://github.com/php/php-src/blob/6828f9d1af7faa18a563d370d84a0182460f5605/Zend/zend_portability.h#L348-L365

Reply:
I've always been hesitant to use `alloca` for the reasons people usually are (i.e. failure when not enough stack?). If you still have to free, it clobbers a lot of the ergonomics of it too...

Reply:
I think technically, VLAs are worse in the "failure is hard to deal with" department than `alloca`. It's really unfortunate C doesn't have similar facilities to C++ in terms of scoped (de)allocations/RAII.